### ISSUE TITLE
Not all attributes are present in module tags with sensible default values

### ISSUE THREAD

**Issue by shs96c** [bazelbuild/bazel#14528]
### Description of the problem / feature request:

When using bazel 5.0.0rc3, declare a module tag similar to:

```starlark
mytag = tag_class(
    attrs = {
        "target": attr.label(),
    },
)
```

If the tag is used in `MODULE.bazel` file _without_ a value (eg. `mytag()`), then in the module's implementation function calls to `mytag.target` will fail, whereas it's expected they return `None`, as happens with empty attributes on repository and regular rules.

**Comment by oquenchil** [comment]
@meteorcloudy

**Comment by shs96c** [comment]
This behaves differently from repository or regular rules. I sincerely hope that this isn't the intended behaviour as it's inconsistent. My expectation would be for these attributes to be set to `None` if they are not specified (as happens with other rule types)

**Comment by Wyverald** [comment]
This certainly isn't intended -- in fact, if you `dir(tag)`, you'll see that the field `target` is supposed to be there. It's just that [the code in `getValue`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/bzlmod/TypeCheckedTag.java;l=113;drc=214a5949291c5a45ff9cc2ad41d6e7508a73167a) returning the default value of "null" causes the field to be considered absent. I'm trying to dig into why this returns `None` for normal rules.

**Comment by Wyverald** [comment]
I've located the root cause -- the fix is in flight.

**Comment by Wyverald** [comment]
Note: We're probably not cherrypicking this fix into 5.0 so please use a rolling release to test bzlmod-related stuff.

**Comment by fmeum** [comment]
> Note: We're probably not cherrypicking this fix into 5.0 so please use a rolling release to test bzlmod-related stuff.

@Wyverald While I agree with the decision not to cherrypick https://github.com/bazelbuild/bazel/pull/14447 into 5.0, I would like to speak out in favor of cherrypicking this fix. bzlmod is picking up traction in the community right now and the release of 5.0 will be when most actively maintained rulesets will start to add support for it. That means that bugs like this, where established patterns unexpectedly don't work in the new setting and have to be worked around in non-optimal ways, will lead to the creation of "legacy code" on day 1 of the availability of this new feature. That can hamper adoption as well as create anti-patterns that will stay with bzlmod for a long time due to backwards compatibility concerns.

**Comment by mattyclarkson** [comment]
I hit an `IllegalArgumentException` when using an `attr.string_dict()` with a `tag_class`. I created a [reproducer project](https://github.com/mattyclarkson/bzlmod-extension-immutable-error).

```
def _rimpl(rctx):
    rctx.file("hey.sh", '#!/bin/sh\nset -eu\necho "${@}"')
    rctx.file("BUILD.bazel", 'exports_files(["hey.sh"], visibility=["//visibility:public"])')

_repo = repository_rule(implementation = _rimpl)

tag = tag_class(
  attrs = {
    "version": attr.string(),
    "test": attr.string_dict(),  # this cannot be read
  },
)

def _impl(mctx):
    for module in mctx.modules:
        for tag in module.tags.test:
            print(tag.test) # 💥 this creates the error
            _repo(name="test-{}".format(tag.version))

test = module_extension(
    implementation = _impl,
    tag_classes = {
        "test": tag,
    },
)
```

End up hitting:

```
Caused by: net.starlark.java.eval.Starlark$UncheckedEvalException: IllegalArgumentException thrown during Starlark evaluation
```

I've [been told](https://bazelbuild.slack.com/archives/C014RARENH0/p1642009847016900?thread_ts=1641550312.005300&cid=C014RARENH0) this issue is the root cause. Hence, could we get this cherry-picked so that we can use `dict`/`list` attributes on `tag_classes` in the `5.0` series?

If we don't cherry-pick the patch: could we enable a _fail-fast_ where at `tag_class` parsing time, any `dict`/`list` attributes bork. Hitting it at runtime (with a hard to grok message) isn't ideal.

**Comment by Wyverald** [comment]
You *can* use `dict`/`list` attributes on `tag_classes`, but you can't _not_ specify those attributes (without the fix, that is).

**Comment by Wyverald** [comment]
re: cherrypicking, after much deliberation with Yun, we've decided to cherrypick the bzlmod-related fixes (~3 commits) into 5.0, given that these are for an explicitly experimental feature and won't affect any other component of Bazel, and are hence very low-risk, *despite* them not strictly being regressions.

This does mean that any breakages resulting from those fixes won't be fixed before 5.0 actually ships. If this does happen (and unrelated bugs are likely to be discovered anyhow), we'll try to fix them in a 5.1 point release.
